### PR TITLE
New block: add "Hand-picked Products" block

### LINKED ISSUE #244: Create a reusable ProductsControl component
Like `ProductCategoryControl`, This will also use the `SearchListControl` behind the scenes. This displays a searchable, selectable list of products.

i3-search-results-2x

This is used by the Hand-picked Products block

### PR DESCRIPTION
**Pull request by ryelle**
Fixes #244, #245 – This PR adds both the ProductsControl shared component, and the Hand-picked Products block which uses that.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader –  _VoiceOver isn't working for me_ 😕 _will test later, but it's using the same patterns as previously tested components_
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

screen shot 2018-12-20 at 11 24 06 am
screen shot 2018-12-20 at 11 23 24 am

Note: this is using the same "columns" slider as the other blocks, rather than the S/M/L column width in the mockup– we can move to the new layout controls once we're not using the shortcode method anymore.

### How to test the changes in this Pull Request:

1. Edit a post/page/etc
2. Try to add a "Hand-picked Products" block
3. Expect: It exists, adding it should bring up the placeholder "edit mode"
4. Your products should load in, search & select some
5. Click Done
6. Expect: The block should flip to a preview, with your products displayed
7. Change the ordering, columns value, or add/remove other products in the sidebar
8. Save the post, view it
9. Expect: The same products, in the same order as the preview.

### CODE REVIEW

**Review by LevinMedia — APPROVED** [review]
Testing well. Thank you!

File: assets/js/handpicked-products.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+			},
+			false
+		);
+		if ( hasChange ) {
+			this.debouncedGetProducts();
+		}
+	}
+
+	getProducts() {
+		if ( ! this.props.attributes.products.length ) {
```
Comment (timmyc) [inline assets/js/handpicked-products.js]: I see `this.props.attributes` is required, but any chance the `.products` key might not be set here?

File: assets/js/handpicked-products.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+			// We've removed all selected products, or products haven't been selected yet.
+			this.setState( { products: [], loaded: true } );
+			return;
+		}
+		apiFetch( {
+			path: addQueryArgs(
+				'/wc-pb/v3/products',
+				getQuery( this.props.attributes, this.props.name )
+			),
+		} )
```
Comment (timmyc) [inline assets/js/handpicked-products.js]: Looks like the tabs are a bit wonky here?

File: assets/js/handpicked-products.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+			</InspectorControls>
+		);
+	}
+
+	renderEditMode() {
+		const { attributes, debouncedSpeak, setAttributes } = this.props;
+		const onDone = () => {
+			setAttributes( { editMode: false } );
+			debouncedSpeak(
+				__( 'Showing product block preview.', 'woo-gutenberg-products-block' )
```
Comment (timmyc) [inline assets/js/handpicked-products.js]: should this say "hand picked product block"?

File: assets/js/handpicked-products.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+	render() {
+		const { setAttributes } = this.props;
+		const { align, columns, editMode } = this.props.attributes;
+		const { loaded, products } = this.state;
+		const classes = [ 'wc-block-products-grid', 'wc-block-handpicked-products' ];
+		if ( columns ) {
+			classes.push( `cols-${ columns }` );
+		}
+		if ( products && ! products.length ) {
```
Comment (timmyc) [inline assets/js/handpicked-products.js]: `products.length` is also checked further down in the render logic - should we create a `hasSelectedProducts` variable to re-use?

File: assets/js/index.js [hunk 5]
```diff
@@ -242,3 +244,81 @@ registerBlockType( 'woocommerce/product-new', {
 		);
 	},
 } );
+
+/**
+ * Register and run the "Products by Category" block.
```
Comment (timmyc) [inline assets/js/index.js]: "Handpicked Products" block.

**Review by timmyc — COMMENTED** [review]
Couple minor comments on the code, but the block is testing out wonderfully for me.

File: assets/js/handpicked-products.js (hunk 2 above)
Comment (timmyc) [reply assets/js/handpicked-products.js]: nevermind, must have been pre-coffee this morning.

**Review by timmyc — APPROVED** [review]
Thanks for the updates - all is looking good!

**Pull request merged**